### DISCUSSION TITLE
Justification Branching Heuristic

### DISCUSSION THREAD

**Discussion by maartenflippo** [cvc5/cvc5#9364] [Q&A] [answered]
Hi,

I am currently familiarizing myself with the implementations of different SMT solvers, and I have a question regarding CVC5, and more specifically its branching heuristic: Is it correct that this heuristic does not make any dynamic decisions based on what happens during search? With dynamic here I mean something akin to how the VSIDS branching heuristic is influenced by literal activity during the search. 

I have read the paper discussing the heuristic, ["A branching heuristic in CVC4 SMT Solver"](https://kshitij.io/articles/cvc4-branching-heuristic.pdf) by Kshitij Bansal, and it is not explicitly mentioned how the formula DAG is traversed recursively. I also looked at the implementation [here](https://github.com/cvc5/cvc5/blob/c42fc9dd0576acb23bd1b4cdb9386e3326c76ec3/src/decision/justification_strategy.cpp) and it leads me to think the traversal of the DAG is done in a fixed order, but perhaps I have overlooked something.

Thank you for your time!

**Comment by ajreynol** [comment, marked as answer]
Yes, you're correct, the justification heuristic used by cvc5 typically just traverses the input formula in a fixed order.

There is an input option `--jh-rlv-order` that adds a coarse activity-based heuristic, where assertions are moved to the front if a conflict is discovered while trying to justify them.  See https://github.com/cvc5/cvc5/blob/c42fc9dd0576acb23bd1b4cdb9386e3326c76ec3/src/decision/assertion_list.h#L33 for some documentation.
This has not made a big difference in most problems, so it is not enabled by default.

We may consider revisiting this in the future.

**Reply by maartenflippo** [reply]
Thanks for the answer. It is interesting that something which makes such a difference in SAT solvers apparently does not have the same benefit for SMT.

**Reply by martin-cs** [reply]
@maartenflippo Yes; this used to confuse the hell out of me.  Having discussed it with various SMT folks, there is an answer which I can believe.  The SAT formulae that are generated by the Boolean skeleton of SMT problems tend to be quite simple.  This is a "way people use SMT solvers" observation rather than anything theoretical.  For example, if you think of the formulae generated by symbolic execution, there are some ITEs and some disjunction but this is mostly 2-SAT.  Consequentially, the main SAT solver in most DPLL(T) systems functions more as a way of marshalling theory lemmas than a major part of the search.  So improvements there don't make that much of a difference to over-all performance.

One could probably do some profiling to investigate this (and, if you did, I suspect the resultant paper and revision of the DPLL(T) architecture would be *fascinating*) but this was enough explanation to keep me happy.

**Reply by maartenflippo** [reply]
That makes sense. I had realized that the boolean skeleton tends to be simple, but I had not connected this fact to the branching heuristic. Thank you for the insight!

**Reply by martin-cs** [reply]
If you want a fun challenge, a theory-aware branching heuristic would be an interesting piece of work.

**Comment by barrettcw** [comment]
To add one note to this thread...Martin is correct that activity-based
heuristics do not matter much for most SMT problems, and of course, this is
just because of the kinds of problems we have - it would be trivial to
construct problems where activity *does* matter but we haven't seen many of
those.  What does matter though, is ignoring irrelevant parts of the
formula.  That is why we go to the trouble of having a "justification"
heuristic.  It does make a big difference on a lot of problems.

We've made various attempts to get the best of both, but as Andy says, most
of the time it doesn't make a big difference.  Incidentally, z3 takes a
different approach, as described here:
https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/tr-2007-140.pdf

Andy has been experimenting with something similar in cvc5, and as I
understand it, it helps sometimes.

-Clark
